### PR TITLE
tools: Wrapper around DuckDuckGo search API

### PR DESCRIPTION
**Pull request by Struki84**
Added a free alternative to google/serpai, wrapper around DuckDuckGo search api. Difference from serpapi, aside from the search engine, is that the search results are in a format of a structured string containing 5 five search results. 

ie. 

```
Title: Leonardo DiCaprio & Camila Morrone's Full Relationship Timeline
Source: duckduckgo.com
Description: Leonardo DiCaprio and Camila Morrone are pros at keeping their relationship under wraps from public scrutiny.
URL: //duckduckgo.com/l/?uddg=https://www.harpersbazaar.com/celebrity/latest/a30841751/leonardo-dicaprio-camila-morrone-relationship-timeline/&rut=89ab942c541611a9bce3d7d0f1b017f5f8a11599476032efa74bbf3131de9804

Title: Who Is Leonardo DiCaprio Dating Now 2022? Victoria Lamas, Gigi Hadid ...
Source: duckduckgo.com
Description: After getting out of very highly publicized relationships, many Leo fans are asking: Who is Leonardo DiCaprio dating now? After launching his career as an actor in the 90s, Leo has been in many...
URL: //duckduckgo.com/l/?uddg=https://stylecaster.com/entertainment/celebrity-news/1348452/leonardo-dicaprio-dating/&rut=8664a8f16ce77d285924fe949a698484ec48cb1f9cf3aa16fbea349e383891cc
```

I wanted to have a little more details about the search to then pass to an LLM for analysis.

### PR Checklist

- [x] Read the [Contributing documentation](https://github.com/tmc/langchaingo/blob/main/CONTRIBUTING.md).
- [x] Read the [Code of conduct documentation](https://github.com/tmc/langchaingo/blob/main/CODE_OF_CONDUCT.md).
- [x] Name your Pull Request title clearly, concisely, and prefixed with the name of primarily affected package you changed according to [Good commit messages](https://go.dev/doc/contribute#commit_messages) (such as `memory: add interfaces for X, Y` or `util: add whizzbang helpers`).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Describes source of new concepts.
- [x] References existing implementations as appropriate.
- [x] Contains test coverage for new functions.
- [x] Passes all [`golangci-lint`](https://golangci-lint.run/) checks.

### CODE REVIEW

File: tools/duckduckgo/ddg.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+func (t Tool) Description() string {
+	return `
+	"A wrapper around DuckDuckGo Search."
+	"Free search alternative to google and serpapi."
+	"Input should be a search query."`
+}
+
+func (t Tool) Call(_ context.Context, input string) (string, error) {
```
Comment (tmc) [inline tools/duckduckgo/ddg.go:32]: the context shouldn't be ignored and should be used in the underlying search.

File: tools/duckduckgo/ddg.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	result, err := t.client.Search(input)
+	if err != nil {
+		if errors.Is(err, internal.ErrNoGoodResult) {
+			return "No good DuckDuckGo Search Results was found", nil
```
Comment (tmc) [inline tools/duckduckgo/ddg.go:36]: should this not return an error?

Comment (FluffyKebab) [reply tools/duckduckgo/ddg.go:36]: If it doesn't return an error the agent will have the possibility to try a different search query or tool.

File: tools/duckduckgo/ddg.go (hunk 1 above)
Comment (Struki84) [reply tools/duckduckgo/ddg.go]: Rewrote the ddg API request to include context, should be ok now...

File: tools/duckduckgo/ddg.go (hunk 2 above)
Comment (Struki84) [reply tools/duckduckgo/ddg.go:39]: Do I need to make a any changes here then?

File: tools/duckduckgo/internal/client.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	return client.formatResults(results), nil
+}
+
+func (client *Client) SetMaxResults(n int) {
+	client.maxResults = n
+}
+
+// formatResults will return a structured string with the results in the format:
```
Comment (tmc) [inline tools/duckduckgo/internal/client.go:94]: this comment isn't attached to the method

**Review by tmc — APPROVED** [review]
LGTM

**Comment by tmc** [comment]
@Struki84 please address these conflicts

**Pull request merged**